File: services/ai-service/src/routers/grammar.py
"""语法学习 Router"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional, List, Any, Dict
from services.grammar_agent import generate_explanation, generate_practice, generate_lesson_practice
from utils.logger import logger

router = APIRouter(prefix="/grammar", tags=["grammar"])


class ExplanationRequest(BaseModel):
    name_zh: str
    name_en: str
    difficulty: str = "intermediate"


class ExplanationResponse(BaseModel):
    explanation: str
    examples: list = []


class PracticeRequest(BaseModel):
    name_zh: str
    name_en: str
    difficulty: str = "intermediate"
    count: int = 10


class PracticeQuestion(BaseModel):
    type: str
    question: str
    options: Optional[List[str]] = None
    answer: str
    explanation: str = ""


class PracticeResponse(BaseModel):
    questions: List[PracticeQuestion]


class LessonPracticeRequest(BaseModel):
    name_zh: str
    name_en: str
    difficulty: str = "intermediate"


class LessonQuestion(BaseModel):
    type: str
    cognitiveLevel: str = "recognition"
    question: str = ""
    options: Optional[List[str]] = None
    answer: str
    answer2: Optional[str] = None
    explanation: str = ""
    errorPart: Optional[str] = None
    correctVersion: Optional[str] = None
    sentence1: Optional[str] = None
    sentence2: Optional[str] = None
    tableData: Optional[Dict[str, Any]] = None
    words: Optional[List[str]] = None
    distractors: Optional[List[str]] = None
    structureHint: Optional[str] = None
    acceptableAnswers: Optional[List[str]] = None
    smartTip: Optional[Dict[str, Any]] = None


class LessonPracticeResponse(BaseModel):
    questions: List[LessonQuestion]


@router.post("/explanation", response_model=ExplanationResponse)
async def get_explanation(request: ExplanationRequest):
    """生成语法点 AI 讲解"""
    try:
        result = await generate_explanation(
            name_zh=request.name_zh,
            name_en=request.name_en,
            difficulty=request.difficulty,
        )
        return result
    except Exception as e:
        logger.error(f"生成语法讲解失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/practice", response_model=PracticeResponse)
async def get_practice(request: PracticeRequest):
    """生成语法练习题"""
    try:
        result = await generate_practice(
            name_zh=request.name_zh,
            name_en=request.name_en,
            difficulty=request.difficulty,
            count=request.count,
        )
        return result
    except Exception as e:
        logger.error(f"生成语法练习题失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/lesson-practice", response_model=LessonPracticeResponse)
async def get_lesson_practice(request: LessonPracticeRequest):
    """生成课程式练习题（8 种题型，按认知层级分组）"""
    try:
        result = await generate_lesson_practice(
            name_zh=request.name_zh,
            name_en=request.name_en,
            difficulty=request.difficulty,
        )
        return result
    except Exception as e:
        logger.error(f"生成课程练习题失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))

File: apps/mobile/src/api/modules/grammar.ts
import { Client } from '../client';
import { ENDPOINTS } from '../endpoints';

// 结构化语法讲解类型
export interface GrammarUsage {
  title: string;
  description?: string;
  exampleEn: string;
  exampleCn: string;
}

export interface GrammarExample {
  en: string;
  cn: string;
  highlight?: string;
}

export interface GrammarCommonError {
  wrong: string;
  correct: string;
  explanation: string;
}

export interface GrammarSections {
  definition: string;
  structure: string;
  usages: GrammarUsage[];
  examples: GrammarExample[];
  commonErrors: GrammarCommonError[];
  tips: string[];
}

export interface StructuredExplanation {
  audioSummary: string;
  sections: GrammarSections;
}

export interface GrammarExplanationData {
  id: string;
  code: string;
  nameZh: string;
  nameEn: string;
  difficulty: string;
  explanation: StructuredExplanation | string;
  audioSummary?: string | null;
  audioUrl?: string | null;
  examples: any[];
}

export interface GrammarCategoryData {
  id: string;
  code: string;
  nameZh: string;
  nameEn: string;
  icon: string;
  color: string;
  level: number;
  pointCount: number;
  learnedCount: number;
}

export type GrammarPointStatus = 'not_started' | 'learning' | 'practiced' | 'mastered';

export interface GrammarPointData {
  id: string;
  code: string;
  nameZh: string;
  nameEn: string;
  difficulty: string;
  status: GrammarPointStatus;
  practiceScore: number | null;
  starRating: number | null;
  bestStarRating: number | null;
}

export interface GrammarCategoryPointsData {
  category: {
    code: string;
    nameZh: string;
    nameEn: string;
    icon: string;
    color: string;
  };
  points: GrammarPointData[];
}

export interface GrammarPracticeQuestion {
  id: string;
  type: 'choice' | 'fill_blank';
  question: string;
  options?: string[];
  answer: string;
  explanation: string;
}

// 课程式练习题型
export type LessonQuestionType =
  | 'choice' | 'fill_blank'
  | 'error_judgment' | 'error_correction'
  | 'dual_blank' | 'table_fill'
  | 'sentence_reorder' | 'word_assembly';

export type CognitiveLevel = 'recognition' | 'understanding' | 'production';

export interface SmartTip {
  rule: string;
  wrong: string;
  correct: string;
  examples: string[];
}

export interface TableBlank {
  row: number;
  col: number;
  answer: string;
  options: string[];
}

export interface TableData {
  headers: string[];
  rows: (string | '___')[][];
  blanks: TableBlank[];
}

export interface LessonQuestion {
  id?: string;
  type: LessonQuestionType;
  cognitiveLevel: CognitiveLevel;
  question: string;
  options?: string[];
  answer: string;
  answer2?: string;
  explanation: string;
  smartTip?: SmartTip;
  errorPart?: string;
  correctVersion?: string;
  sentence1?: string;
  sentence2?: string;
  tableData?: TableData;
  words?: string[];
  distractors?: string[];
  chineseTranslation?: string;
  structureHint?: string;
  acceptableAnswers?: string[];
}

export interface GrammarLessonData {
  point: {
    id: string;
    code: string;
    nameZh: string;
    nameEn: string;
    difficulty: string;
  };
  explanation: StructuredExplanation | string | null;
  questions: LessonQuestion[];
}

declare module '../client' {
  interface Client {
    getGrammarCategories(): Promise<{
      success: boolean; data: GrammarCategoryData[];
    }>;
    getGrammarCategoryPoints(categoryCode: string): Promise<{
      success: boolean; data: GrammarCategoryPointsData;
    }>;
    getGrammarExplanation(pointCode: string): Promise<{
      success: boolean; data: GrammarExplanationData;
    }>;
    getGrammarPractice(pointCode: string, count?: number): Promise<{
      success: boolean; data: GrammarPracticeQuestion[];
    }>;
    submitGrammarPractice(pointId: string, result: {
      score: number;
      totalCount: number;
      correctCount: number;
    }): Promise<{ success: boolean }>;
    getGrammarLesson(pointCode: string): Promise<{
      success: boolean; data: GrammarLessonData;
    }>;
    submitGrammarLesson(pointId: string, result: {
      score: number;
      totalCount: number;
      correctCount: number;
      starRating: number;
      phaseResults?: { phase: string; correctRate: number }[];
    }): Promise<{ success: boolean }>;
  }
}

Client.prototype.getGrammarCategories = async function() {
  return this.api.get(ENDPOINTS.GRAMMAR_CATEGORIES);
};

Client.prototype.getGrammarCategoryPoints = async function(categoryCode: string) {
  return this.api.get(`${ENDPOINTS.GRAMMAR_CATEGORIES}/${categoryCode}/points`);
};

Client.prototype.getGrammarExplanation = async function(pointCode: string) {
  return this.api.get(`${ENDPOINTS.GRAMMAR_POINTS}/${pointCode}/explanation`);
};

Client.prototype.getGrammarPractice = async function(pointCode: string, count = 10) {
  return this.api.post(`${ENDPOINTS.GRAMMAR_POINTS}/${pointCode}/practice`, { count });
};

Client.prototype.submitGrammarPractice = async function(pointId: string, result) {
  return this.api.post(ENDPOINTS.GRAMMAR_SUBMIT, { pointId, ...result });
};

Client.prototype.getGrammarLesson = async function(pointCode: string) {
  return this.api.post(`${ENDPOINTS.GRAMMAR_POINTS}/${pointCode}/lesson`);
};

Client.prototype.submitGrammarLesson = async function(pointId: string, result) {
  return this.api.post(ENDPOINTS.GRAMMAR_LESSON_SUBMIT, { pointId, ...result });
};
